tracing/tracing.go

```diff
@@ -0,0 +1,124 @@
+
+	srvconfig "github.com/containerd/containerd/services/server/config"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc"
+)
+
+// InitOpenTelemetry read config and initializes otel middleware, sets the exporter
+// propagator and global tracer provider.
+func InitOpenTelemetry(config *srvconfig.Config) (func(), error) {
+	ctx := context.Background()
+
+	// Check if tracing is configured
+	if config.OpenTelemetry == *new(srvconfig.OpenTelemetryConfig) {
+		logrus.Info("OpenTelemetry configuration is not found in config, tracing will be disabled")
+		return nil, nil
+	}
+
+	// Validate configuration
+	err := config.OpenTelemetry.Validate()
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid open telemetry configuration")
+	}
+
+	res, err := resource.New(ctx,
+		resource.WithAttributes(
+			// Service name used to displace traces in backends
+			semconv.ServiceNameKey.String("containerd"),
+		),
+	)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create resource")
+	}
+
+	// Configure OTLP trace exporter and set it up to connect to OpenTelemetry collector
+	// running on a local host.
+	ctrdTraceExporter, err := otlptracegrpc.New(ctx,
+		otlptracegrpc.WithEndpoint(config.OpenTelemetry.ExporterEndPoint),
+		otlptracegrpc.WithDialOption(grpc.WithBlock()),
+	)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create trace exporter")
+	}
+
+	// Register the trace exporter with a TracerProvider, using a batch span
+	// process to aggregate spans before export.
+	ctrdBatchSpanProcessor := sdktrace.NewBatchSpanProcessor(ctrdTraceExporter)
+	ctrdTracerProvider := sdktrace.NewTracerProvider(
+		// We use TraceIDRatioBased sampling. Ration read from config translated into following
+		// if sampleing ration < 0 it is interpreted as 0. If ration >= 1, it will always sample.
+		sdktrace.WithSampler(sdktrace.TraceIDRatioBased(config.OpenTelemetry.TraceSamplingRatio)),
+		sdktrace.WithResource(res),
+		sdktrace.WithSpanProcessor(ctrdBatchSpanProcessor),
+	)
+	otel.SetTracerProvider(ctrdTracerProvider)
+
+	// set global propagator to tracecontext
+	otel.SetTextMapPropagator(propagation.TraceContext{})
+
+	return func() {
+		// Shutdown will flush any remaining spans and shut down the exporter
+		err := ctrdTracerProvider.Shutdown(ctx)
+		if err != nil {
+			logrus.Errorf("failed to shutdown TracerProvider: %v", err)
+		}
+
+	}, nil
+}
+
+// StartSpan starts child span in a context.
+func StartSpan(ctx context.Context, opName string, opts ...trace.SpanStartOption) (trace.Span, context.Context) {
+	parentSpan := trace.SpanFromContext(ctx)
+	tracer := trace.NewNoopTracerProvider().Tracer("")
```

Comment:
How about putting this line in `else`? If the parent span is valid, this value would be immediately overridden by the next line.

Reply:
Actually that is the intention here. If parent span is valid we get the tracer from parent, Otherwise keep using the one already initialized.